src/storage/secondary/column/char_column_builder.rs

```diff
@@ -130,6 +144,13 @@ impl ColumnBuilder<Utf8Array> for CharColumnBuilder {
                         nullable
                     ),
                 }
+
+                if let Some(to_be_appended) = iter.peek() {
+                    if self.options.record_first_key && to_be_appended.is_some() {
+                        self.is_first_key_null = false;
+                        self.first_key = to_be_appended.unwrap().as_bytes().to_vec();
+                    }
```

Comment:
It seems that once `is_first_key_null` is set to `false`, it will never be set back to `true`?

We can define `self.first_key` as `Option<Vec<u8>>`, so as to make the logic clearer.

```rust
if self.options.record_first_key {
   self.first_key = to_be_appended.map(|x| x.as_bytes().to_vec());
}
```

And we can modify `finish_block` to accept `first_key: Option<Vec<u8>>`. Inside the block builder, we match `first_key`. If null, set `is_first_key_null` to true. Otherwise, set `first_key` to the inner value.

Reply:
> Still some small issues. Thanks again for your contribution!

Thanks also for your review too! 😭

Reply:
> It seems that once `is_first_key_null` is set to `false`, it will never be set back to `true`?
> 
> We can define `self.first_key` as `Option<Vec<u8>>`, so as to make the logic clearer.
> 
> ```rust
> if self.options.record_first_key {
>    self.first_key = to_be_appended.map(|x| x.as_bytes().to_vec());
> }
> ```
> 
> And we can modify `finish_block` to accept `first_key: Option<Vec<u8>>`. Inside the block builder, we match `first_key`. If null, set `is_first_key_null` to true. Otherwise, set `first_key` to the inner value.

Maybe `finish_block()` will be like this?

```rust
    pub fn finish_block(
        &mut self,
        block_type: BlockType,
        column_data: &mut Vec<u8>,
        block_data: &mut Vec<u8>,
        stats: Vec<BlockStatistics>,
        first_key: &mut Option<Vec<u8>>,
    ) {
        self.indexes.push(BlockIndex {
            offset: column_data.len() as u64,
            length: block_data.len() as u64 + BLOCK_HEADER_SIZE as u64,
            first_rowid: self.last_row_count as u32,
            row_count: (self.row_count - self.last_row_count) as u32,
            /// TODO(chi): support sort key
            first_key: match first_key {
                Some(x) => x.drain(..).collect(),
                None => Default::default(),
            },
            stats,
            is_first_key_null: first_key.is_none(),
        });
        ...
```

Reply:
Exactly! And I’d prefer to have Option<Vec<u8>> as parameter instead of &mut. As block is typically large, this re-allocation is acceptable.